src/Event/Emitter.cpp

```diff
@@ -38,36 +39,40 @@ Emitter::Emitter()
 
 Emitter::~Emitter()
 {
+    Dispatcher::getInstance()->removeEventHandler(this);
 }
 
-void Emitter::addEventHandler(const std::string& eventName, std::function<void(Event*)> handler)
+void Emitter::addEventHandler(const std::string& eventName, Emitter::Handler handler)
 {
-    if (_eventHandlers.find(eventName) == _eventHandlers.end())
-    {
-        std::vector<std::function<void(Event*)>> vector;
-        _eventHandlers.insert(std::make_pair(eventName, vector));
-    }
-
-    _eventHandlers.at(eventName).push_back(handler);
+    _eventHandlers[eventName].push_back(handler);
 }
 
-void Emitter::emitEvent(Event* event)
+void Emitter::emitEvent(std::unique_ptr<Event> event)
 {
     if (_eventHandlers.find(event->name()) == _eventHandlers.end()) return;
+
+    Dispatcher::getInstance()->postEventHandler(this, std::move(event));
+}
+
+void Emitter::processEvent(std::unique_ptr<Event> event)
+{
```

Comment:
The previous implementation had a capability to group handlers by event name.
I just split it into two phases - delayed post and sync processing.
Do you think that now we should name it like EventSource or something?